# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [2]:
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [3]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [4]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [6]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [5]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [7]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [8]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [9]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [10]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [11]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [12]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [13]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [14]:
df['Season'].value_counts()

2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: Season, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [15]:
Useasons = df['Season'].nunique()
print("Number of unique seasons:", Useasons)

Number of unique seasons: 32


**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [16]:
Team_Wins = df['Wteam'].value_counts().idxmax()
print("Team with the most wins:", Team_Wins)

Team with the most wins: 1181


# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [17]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [19]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [20]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [21]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [22]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [23]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [24]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [25]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [26]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [27]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [30]:
# Sort DataFrame by season
df_sorted_by_season = df.sort_values('Season')
print("Sorted by season:")
display(df_sorted_by_season.head())

# Sort DataFrame by winning team
df_sorted_by_winning_team = df.sort_values('Wteam')
print("Sorted by winning team:")
display(df_sorted_by_winning_team.head())

# Sort DataFrame by winning score
df_sorted_by_winning_score = df.sort_values('Wscore')
print("Sorted by winning score:")
display(df_sorted_by_winning_score.head())


Sorted by season:


Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0       1985      20   1228      81   1328      64    N      0
2484    1985      99   1275      73   1132      63    H      0
2485    1985      99   1306      62   1171      59    H      0
2486    1985      99   1318      99   1145      91    H      0
2487    1985      99   1337      75   1109      56    H      0

Sorted by winning team:


Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
140642    2016      25   1101      72   1197      62    N      0
136740    2015      62   1101      87   1146      70    H      0
143283    2016      89   1101      80   1249      71    H      0
143476    2016      92   1101      79   1223      72    H      0
141972    2016      61   1101      62   1146      61    A      0

Sorted by winning score:


Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
128582    2013     116   1264      34   1193      31    H      0
69336     2001     126   1206      35   1423      33    N      0
5364      1986      77   1229      35   1166      34    A      0
20246     1990      47   1374      36   1426      31    H      0
118590    2011     130   1336      36   1458      33    N      0

In [31]:
df.iloc[100:201][["Season","Wteam","Wscore"]]

Season  Wteam  Wscore
100    1985   1409      86
101    1985   1444      90
102    1985   1451      89
103    1985   1139      78
104    1985   1154      78
..      ...    ...     ...
196    1985   1203      83
197    1985   1222      73
198    1985   1227      76
199    1985   1228      73
200    1985   1229      54

[101 rows x 3 columns]

**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [32]:
# Find the game with the highest winning score in each subset
max_winning_score_season_sorted_by_season = df_sorted_by_season[df_sorted_by_season['Wscore'] == df_sorted_by_season['Wscore'].max()]['Season'].values[0]
max_winning_score_team_sorted_by_season = df_sorted_by_season[df_sorted_by_season['Wscore'] == df_sorted_by_season['Wscore'].max()]['Wteam'].values[0]

max_winning_score_season_sorted_by_winning_team = df_sorted_by_winning_team[df_sorted_by_winning_team['Wscore'] == df_sorted_by_winning_team['Wscore'].max()]['Season'].values[0]
max_winning_score_team_sorted_by_winning_team = df_sorted_by_winning_team[df_sorted_by_winning_team['Wscore'] == df_sorted_by_winning_team['Wscore'].max()]['Wteam'].values[0]

max_winning_score_season_sorted_by_winning_score = df_sorted_by_winning_score[df_sorted_by_winning_score['Wscore'] == df_sorted_by_winning_score['Wscore'].max()]['Season'].values[0]
max_winning_score_team_sorted_by_winning_score = df_sorted_by_winning_score[df_sorted_by_winning_score['Wscore'] == df_sorted_by_winning_score['Wscore'].max()]['Wteam'].values[0]

# Print the results
print("From the subset sorted by season:")
print("Season:", max_winning_score_season_sorted_by_season)
print("Winning team:", max_winning_score_team_sorted_by_season)

print("\nFrom the subset sorted by winning team:")
print("Season:", max_winning_score_season_sorted_by_winning_team)
print("Winning team:", max_winning_score_team_sorted_by_winning_team)

print("\nFrom the subset sorted by winning score:")
print("Season:", max_winning_score_season_sorted_by_winning_score)
print("Winning team:", max_winning_score_team_sorted_by_winning_score)


From the subset sorted by season:
Season: 1991
Winning team: 1258

From the subset sorted by winning team:
Season: 1991
Winning team: 1258

From the subset sorted by winning score:
Season: 1991
Winning team: 1258


Since the three dataframes provide the same result, it attests to the fact that the dataframes offer the same accurate information.

# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [33]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [34]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [36]:
# Create a new column 'ScoreDifference'
df['ScoreDifference'] = abs(df['Wscore'] - df['Lscore'])

# Calculate the average 'ScoreDifference' for all games
average_score_difference = df['ScoreDifference'].mean()

# Filter the DataFrame based on the condition
filtered_df = df[df['ScoreDifference'] > average_score_difference]

# Print the filtered DataFrame using display()
print("Filtered DataFrame:")
print(filtered_df)
display(filtered_df)


Filtered DataFrame:
        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
3         1985      25   1165      70   1432      54    H      0   
6         1985      25   1228      64   1226      44    N      0   
8         1985      25   1260      98   1133      80    H      0   
10        1985      25   1307     103   1288      71    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145280    2016     131   1401      71   1261      38    N      0   
145282    2016     131   1433      76   1172      54    N      0   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference  
0                    17  
3                    16  
6                    20  
8                    18  
10                   32  
...    

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
3         1985      25   1165      70   1432      54    H      0   
6         1985      25   1228      64   1226      44    N      0   
8         1985      25   1260      98   1133      80    H      0   
10        1985      25   1307     103   1288      71    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145280    2016     131   1401      71   1261      38    N      0   
145282    2016     131   1433      76   1172      54    N      0   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference  
0                    17  
3                    16  
6                    20  
8                    18  
10                   32  
...                 ...  
145280               33  
145282               22  
145284               20  
145285               14  
145288               13  

[57227 rows x 9 columns]

**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [37]:
# Find the game with the highest 'ScoreDifference' in the filtered DataFrame
max_score_difference_game = filtered_df[filtered_df['ScoreDifference'] == filtered_df['ScoreDifference'].max()]

# Get the season and teams involved in the game
season = max_score_difference_game['Season'].values[0]
winning_team = max_score_difference_game['Wteam'].values[0]
losing_team = max_score_difference_game['Lteam'].values[0]

# Print the results
print("Season:", season)
print("Winning Team:", winning_team)
print("Losing Team:", losing_team)


Season: 1996
Winning Team: 1409
Losing Team: 1341


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [38]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [39]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: Wloc, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [40]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [41]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [42]:
df.groupby('Season').mean("Wscore")

Daynum        Wteam     Wscore        Lteam     Lscore     Numot  \
Season                                                                        
1985    81.817233  1287.701097  74.723040  1281.085898  63.860316  0.007760   
1986    81.445678  1285.545863  74.813640  1282.876024  63.633888  0.002115   
1987    82.907535  1284.852107  77.993870  1283.150192  66.770626  0.008429   
1988    79.438938  1284.863970  79.773704  1284.150695  67.697092  0.003793   
1989    79.123854  1283.772851  81.728511  1285.397325  69.530840  0.005697   
1990    79.195303  1282.814339  80.846477  1284.844499  68.833622  0.006180   
1991    78.725928  1287.633519  82.733932  1282.486054  70.057240  0.007519   
1992    76.514660  1285.796947  79.992004  1284.502060  67.144657  0.003150   
1993    81.057258  1286.536414  79.508790  1285.797589  67.301858  0.004771   
1994    80.279557  1287.935468  81.160345  1286.099261  68.561330  0.005665   
1995    79.828550  1291.149129  80.096149  1280.886191  67.394898  0.008094   
1996    79.872392  1288.701601  78.205483  1284.702572  65.737506  0.007035   
1997    80.003369  1286.539832  76.300842  1285.100602  63.742238  0.007942   
1998    78.013679  1288.089753  77.229902  1284.805136  64.612911  0.005040   
1999    76.980341  1286.234249  76.043581  1284.950024  63.427996  0.008053   
2000    75.169285  1287.418677  76.310467  1281.264882  63.870768  0.066608   
2001    75.721289  1285.600627  77.088650  1282.611372  64.833221  0.073651   
2002    75.277497  1286.658836  76.963557  1281.581120  64.614270  0.065642   
2003    74.808709  1285.591854  75.795061  1281.750867  63.756932  0.061308   
2004    75.335156  1285.931306  74.971122  1280.900678  62.990155  0.071319   
2005    75.042567  1286.067166  74.693690  1282.088770  62.635294  0.066524   
2006    74.712634  1283.769182  74.587555  1285.425478  62.666597  0.073996   
2007    71.676780  1286.476700  74.829467  1283.896887  62.761650  0.067618   
2008    68.392020  1288.741817  74.962425  1281.907418  62.755375  0.072439   
2009    70.963612  1287.680511  74.033721  1281.312250  61.980377  0.071633   
2010    70.741212  1289.607638  74.652480  1279.997910  62.484325  0.063272   
2011    70.881243  1285.244758  74.374190  1283.078727  62.470644  0.077964   
2012    70.745479  1286.596802  73.379973  1283.055778  61.415953  0.070055   
2013    68.058647  1284.684211  72.840789  1285.594361  60.908083  0.072744   
2014    68.882507  1288.116561  76.101268  1280.580940  64.452070  0.082991   
2015    70.742249  1288.073963  72.680799  1279.929399  61.007845  0.079380   
2016    70.922332  1289.136711  78.295958  1279.221270  66.327063  0.078413   

        ScoreDifference  
Season                   
1985          10.862724  
1986          11.179752  
1987          11.223244  
1988          12.076612  
1989          12.197672  
1990          12.012855  
1991          12.676692  
1992          12.847347  
1993          12.206931  
1994          12.599015  
1995          12.701251  
1996          12.467977  
1997          12.558604  
1998          12.616991  
1999          12.615585  
2000          12.439699  
2001          12.255429  
2002          12.349286  
2003          12.038128  
2004          11.980967  
2005          12.058396  
2006          11.920959  
2007          12.067817  
2008          12.207050  
2009          12.053343  
2010          12.168155  
2011          11.903546  
2012          11.964021  
2013          11.932707  
2014          11.649198  
2015          11.672955  
2016          11.968896

**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [43]:
df.groupby('Wteam').max('Wscore')

Season  Daynum  Wscore  Lteam  Lscore  Numot  ScoreDifference
Wteam                                                               
1101     2016     120      95   1368      84      0               17
1102     2016     129     111   1464      94      3               45
1103     2016     131     109   1464     104      2               58
1104     2016     132     114   1464     101      3               57
1105     2016     132     114   1445      93      2               37
...       ...     ...     ...    ...     ...    ...              ...
1460     2016     131     136   1464     113      2               66
1461     2016     132     112   1457      98      4               53
1462     2016     131     125   1464     113      3               57
1463     2016     131     105   1447      96      2               39
1464     2016     126     115   1460     101      2               34

[364 rows x 7 columns]

**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [45]:
# Group the DataFrame by season and winning team
grouped_df = df.groupby(['Season', 'Wteam'])

# Find the team with the highest average winning score for each season
highest_avg_score_per_season = grouped_df['Wscore'].mean().groupby('Season').idxmax()

# Create an empty list to store the result rows
result_rows = []

# Populate the list with the result rows
for season, team_index in highest_avg_score_per_season.items():
    winning_team = team_index[1]
    avg_score = grouped_df.get_group(team_index)['Wscore'].mean()
    result_rows.append({'Season': season, 'Winning Team': winning_team, 'Average Winning Score': avg_score})

# Create a new DataFrame with the results
result_df = pd.DataFrame(result_rows)

# Print the resulting DataFrame
print("Teams with the highest average winning score for each season:")
display(result_df)


Teams with the highest average winning score for each season:


Season  Winning Team  Average Winning Score
0     1985          1328              92.800000
1     1986          1109              91.200000
2     1987          1380              95.875000
3     1988          1258             111.750000
4     1989          1258             117.315789
5     1990          1258             126.347826
6     1991          1380             112.312500
7     1992          1380              99.642857
8     1993          1380             101.875000
9     1994          1380             106.583333
10    1995          1206             102.833333
11    1996          1206             103.600000
12    1997          1254              94.142857
13    1998          1395             101.153846
14    1999          1317              95.750000
15    2000          1395              95.647059
16    2001          1395              99.315789
17    2002          1242              92.888889
18    2003          1395              88.222222
19    2004          1194              92.142857
20    2005          1311              91.333333
21    2006          1190              89.533333
22    2007          1440              99.400000
23    2008          1440              94.500000
24    2009          1377              96.200000
25    2010          1440              95.000000
26    2011          1449              90.130435
27    2012          1370              87.400000
28    2013          1377              86.500000
29    2014          1322              92.600000
30    2015          1322              90.733333
31    2016          1146              97.142857

**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [47]:
# Group the DataFrame by season and winning team
grouped_df = df.groupby(['Season', 'Wteam'])

# Count the number of wins for each team in each season
wins_per_team_per_season = grouped_df.size().reset_index(name='Wins')

# Print the resulting DataFrame
print("Number of wins for each team in each season:")
display(wins_per_team_per_season)


Number of wins for each team in each season:


Season  Wteam  Wins
0        1985   1102     5
1        1985   1103     9
2        1985   1104    21
3        1985   1106    10
4        1985   1108    19
...       ...    ...   ...
10167    2016   1460    20
10168    2016   1461    12
10169    2016   1462    27
10170    2016   1463    21
10171    2016   1464     9

[10172 rows x 3 columns]

**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [48]:
# Group the DataFrame by season and winning team
grouped_df = df.groupby(['Season', 'Wteam'])

# Count the number of wins for each team in each season
wins_per_team_per_season = grouped_df.size().reset_index(name='Wins')

# Find the team with the most wins for each season
teams_with_most_wins_per_season = wins_per_team_per_season.groupby('Season')['Wins'].idxmax()
teams_with_most_wins_per_season = wins_per_team_per_season.loc[teams_with_most_wins_per_season]

# Print the resulting DataFrame
print("Team with the most wins for each season:")
display(teams_with_most_wins_per_season)


Team with the most wins for each season:


Season  Wteam  Wins
217     1985   1385    27
342     1986   1181    32
818     1987   1424    33
861     1988   1112    31
1323    1989   1328    28
1551    1990   1247    29
1739    1991   1116    30
2088    1992   1181    28
2423    1993   1231    28
2665    1994   1163    27
2928    1995   1116    26
3357    1996   1269    31
3638    1997   1242    31
3943    1998   1242    33
4201    1999   1181    32
4709    2000   1409    29
4829    2001   1181    29
5121    2002   1153    30
5456    2003   1166    29
5981    2004   1390    29
6165    2005   1228    32
6452    2006   1181    30
6841    2007   1242    30
7209    2008   1272    33
7552    2009   1272    31
7870    2010   1242    32
8217    2011   1242    32
8565    2012   1246    32
8878    2013   1211    30
9462    2014   1455    33
9609    2015   1246    34
9955    2016   1242    29

**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [50]:
# Group the DataFrame by losing team
grouped_losing_df = df.groupby('Lteam')

# Calculate the average losing score for each team across all seasons
average_losing_score = grouped_losing_df['Lscore'].mean()

# Print the resulting DataFrame
print("Average losing score for each team across all seasons:")
display(average_losing_score)


Average losing score for each team across all seasons:


Lteam
1101    60.586207
1102    59.201507
1103    64.117347
1104    64.374317
1105    61.675373
          ...    
1460    63.082915
1461    64.661972
1462    68.216117
1463    60.829213
1464    65.187063
Name: Lscore, Length: 364, dtype: float64

I am creating a dataframe that is easier to comprehend by placing the average winning score and losing score for different teams side by side.

In [52]:
# Group the DataFrame by losing team
grouped_losing_df = df.groupby('Lteam')

# Calculate the average losing score for each team across all seasons
average_losing_score = grouped_losing_df['Lscore'].mean()

# Group the DataFrame by winning team and calculate the average winning score for each team across all seasons
average_winning_score = df.groupby('Wteam')['Wscore'].mean()

# Merge the average losing score and average winning score dataframes
merged_scores_df = average_losing_score.to_frame().merge(average_winning_score.to_frame(), left_index=True, right_index=True)

# Rename the columns
merged_scores_df.columns = ['Average Losing Score', 'Average Winning Score']

# Print the merged DataFrame
print("Average losing score and average winning score for each team:")
display(merged_scores_df)


Average losing score and average winning score for each team:


Average Losing Score  Average Winning Score
Lteam                                             
1101              60.586207              78.111111
1102              59.201507              69.893204
1103              64.117347              75.839768
1104              64.374317              75.825944
1105              61.675373              74.960894
...                     ...                    ...
1460              63.082915              75.531469
1461              64.661972              75.170082
1462              68.216117              79.906021
1463              60.829213              71.720102
1464              65.187063              73.926056

[364 rows x 2 columns]

In [54]:
# Filter the merged DataFrame to include only the teams with higher average Lscore than Wscore
filtered_teams = merged_scores_df[merged_scores_df['Average Losing Score'] > merged_scores_df['Average Winning Score']]

# Print the filtered teams
print("Teams with a higher average Lscore than Wscore:")
display(filtered_teams)


Teams with a higher average Lscore than Wscore:


Empty DataFrame
Columns: [Average Losing Score, Average Winning Score]
Index: []

The conclusion must be that there are no teams that have an average Lscore higher than Wscore.

# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [55]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [57]:
# Create a new column 'HighScoringGame'
df['HighScoringGame'] = df['Wscore'].apply(lambda x: 'Yes' if x > 100 else 'No')

# Print the updated DataFrame
print("Updated DataFrame with 'HighScoringGame' column:")
display(df)


Updated DataFrame with 'HighScoringGame' column:


Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
1         1985      25   1106      77   1354      70    H      0   
2         1985      25   1112      63   1223      56    H      0   
3         1985      25   1165      70   1432      54    H      0   
4         1985      25   1192      86   1447      74    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145286    2016     132   1246      82   1401      77    N      1   
145287    2016     132   1277      66   1345      62    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference HighScoringGame  
0                    17              No  
1                     7              No  
2                     7              No  
3                    16              No  
4                    12              No  
...                 ...             ...  
145284               20              No  
145285               14              No  
145286                5              No  
145287                4              No  
145288               13              No  

[145289 rows x 10 columns]

Since I couldn't identify any 'Yes' values in the above DataFrame, I am trying to create one that extracts the results with 'Yes' just to verify if the code is correct.

In [58]:
# Filter the DataFrame to include only rows with 'HighScoringGame' as 'Yes'
high_scoring_games_df = df[df['HighScoringGame'] == 'Yes']

# Print the filtered DataFrame
print("DataFrame with 'HighScoringGame' as 'Yes':")
display(high_scoring_games_df)


DataFrame with 'HighScoringGame' as 'Yes':


Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
10        1985      25   1307     103   1288      71    H      0   
25        1985      26   1196     106   1416      55    H      0   
35        1985      26   1286     109   1186      64    H      0   
42        1985      26   1375     104   1126      68    H      0   
50        1985      26   1439     102   1330      76    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
144999    2016     124   1267     108   1379     106    H      1   
145135    2016     128   1266     101   1385      93    N      0   
145148    2016     128   1424     108   1102     102    N      3   
145218    2016     130   1163     104   1153      97    N      4   
145245    2016     130   1372     104   1223      68    N      0   

        ScoreDifference HighScoringGame  
10                   32             Yes  
25                   51             Yes  
35                   45             Yes  
42                   36             Yes  
50                   26             Yes  
...                 ...             ...  
144999                2             Yes  
145135                8             Yes  
145148                6             Yes  
145218                7             Yes  
145245               36             Yes  

[4880 rows x 10 columns]

**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [60]:
# Initialize a dictionary to keep track of the number of games for each team
team_games = {}

# Iterate over the rows of the DataFrame
for index, row in df.iterrows():
    # Get the winning team and losing team for the current row
    winning_team = row['Wteam']
    losing_team = row['Lteam']
    
    # Update the number of games for the winning team
    if winning_team in team_games:
        team_games[winning_team] += 1
    else:
        team_games[winning_team] = 1
    
    # Update the number of games for the losing team
    if losing_team in team_games:
        team_games[losing_team] += 1
    else:
        team_games[losing_team] = 1

# Print the total number of games played by each team
print("Total number of games played by each team:")
for team, games in team_games.items():
    print(f"{team}: {games} games")


Total number of games played by each team:
1228: 992 games
1328: 968 games
1106: 855 games
1354: 906 games
1112: 981 games
1223: 363 games
1165: 833 games
1432: 69 games
1192: 908 games
1447: 903 games
1218: 931 games
1337: 922 games
1226: 847 games
1242: 993 games
1268: 969 games
1260: 914 games
1133: 949 games
1305: 922 games
1424: 974 games
1307: 969 games
1288: 925 games
1344: 951 games
1438: 952 games
1374: 916 games
1411: 903 games
1412: 962 games
1397: 963 games
1417: 966 games
1225: 880 games
1116: 980 games
1368: 808 games
1120: 936 games
1391: 879 games
1135: 847 games
1306: 898 games
1143: 947 games
1388: 897 games
1153: 970 games
1184: 863 games
1159: 887 games
1171: 829 games
1216: 930 games
1173: 960 games
1134: 200 games
1177: 942 games
1296: 879 games
1193: 942 games
1265: 934 games
1196: 981 games
1416: 881 games
1206: 938 games
1137: 912 games
1210: 972 games
1149: 824 games
1211: 921 games
1102: 840 games
1234: 968 games
1114: 910 games
1332: 927 games
1243: 927 game

**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [65]:
# Create a dictionary to store the highest score difference for each season
highest_score_diff_per_season = {}

# Iterate over the rows of the DataFrame
for index, row in df.iterrows():
    season = row['Season']
    score_diff = row['Wscore'] - row['Lscore']
    
    # Check if the season already exists in the dictionary
    if season in highest_score_diff_per_season:
        # If the current score difference is higher, update the value in the dictionary
        if score_diff > highest_score_diff_per_season[season]['score_diff']:
            highest_score_diff_per_season[season] = {
                'score_diff': score_diff,
                'game': index
            }
    else:
        # If the season is not in the dictionary, initialize it with the current score difference
        highest_score_diff_per_season[season] = {
            'score_diff': score_diff,
            'game': index
        }


# Print the game with the highest score difference for each season
for season, result in highest_score_diff_per_season.items():
    game_index = result['game']
    score_diff = result['score_diff']
    print(f"Season: {season}, Game with highest score difference: {game_index}, Score Difference: {score_diff}")


Season: 1985, Game with highest score difference: 236, Score Difference: 60
Season: 1986, Game with highest score difference: 4731, Score Difference: 84
Season: 1987, Game with highest score difference: 8240, Score Difference: 73
Season: 1988, Game with highest score difference: 12046, Score Difference: 68
Season: 1989, Game with highest score difference: 16677, Score Difference: 70
Season: 1990, Game with highest score difference: 19502, Score Difference: 76
Season: 1991, Game with highest score difference: 25161, Score Difference: 68
Season: 1992, Game with highest score difference: 27997, Score Difference: 82
Season: 1993, Game with highest score difference: 33858, Score Difference: 81
Season: 1994, Game with highest score difference: 36404, Score Difference: 69
Season: 1995, Game with highest score difference: 39858, Score Difference: 74
Season: 1996, Game with highest score difference: 44653, Score Difference: 91
Season: 1997, Game with highest score difference: 49033, Score Diffe

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [64]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [66]:
# Calculate the count of games for each team using value_counts() on the winning team and losing team columns separately
games_played_winning = df['Wteam'].value_counts()
games_played_losing = df['Lteam'].value_counts()

# Add the two Series together to get the total number of games played by each team
games_played = games_played_winning.add(games_played_losing, fill_value=0)

# Print the number of games played by each team
for team, count in games_played.items():
    print(f"Team: {team}, Games Played: {count}")

Team: 1101, Games Played: 76
Team: 1102, Games Played: 840
Team: 1103, Games Played: 910
Team: 1104, Games Played: 975
Team: 1105, Games Played: 447
Team: 1106, Games Played: 855
Team: 1107, Games Played: 512
Team: 1108, Games Played: 866
Team: 1109, Games Played: 169
Team: 1110, Games Played: 910
Team: 1111, Games Played: 876
Team: 1112, Games Played: 981
Team: 1113, Games Played: 937
Team: 1114, Games Played: 910
Team: 1115, Games Played: 520
Team: 1116, Games Played: 980
Team: 1117, Games Played: 891
Team: 1118, Games Played: 27
Team: 1119, Games Played: 868
Team: 1120, Games Played: 936
Team: 1121, Games Played: 175
Team: 1122, Games Played: 903
Team: 1123, Games Played: 906
Team: 1124, Games Played: 906
Team: 1125, Games Played: 506
Team: 1126, Games Played: 906
Team: 1127, Games Played: 442
Team: 1128, Games Played: 106
Team: 1129, Games Played: 907
Team: 1130, Games Played: 951
Team: 1131, Games Played: 945
Team: 1132, Games Played: 895
Team: 1133, Games Played: 949
Team: 1134, 

**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [67]:
# Create a new column 'ScoreDifference' using vectorized subtraction
df['ScoreDifference'] = df['Wscore'] - df['Lscore']

# Use groupby() and idxmax() to find the game with the highest score difference for each season
highest_score_diff_games = df.groupby('Season')['ScoreDifference'].idxmax()

# Retrieve the corresponding game information
highest_score_diff_info = df.loc[highest_score_diff_games, ['Season', 'Wteam', 'Lteam', 'Wscore', 'Lscore', 'ScoreDifference']]

# Print the game with the highest score difference for each season
print("Game with the highest score difference for each season:")
for _, game_info in highest_score_diff_info.iterrows():
    season = game_info['Season']
    winning_team = game_info['Wteam']
    losing_team = game_info['Lteam']
    winning_score = game_info['Wscore']
    losing_score = game_info['Lscore']
    score_difference = game_info['ScoreDifference']
    print(f"Season {season}: Highest score difference = {score_difference}, Game: {winning_team} ({winning_score}) vs {losing_team} ({losing_score})")


Game with the highest score difference for each season:
Season 1985: Highest score difference = 60, Game: 1361 (128) vs 1288 (68)
Season 1986: Highest score difference = 84, Game: 1314 (129) vs 1264 (45)
Season 1987: Highest score difference = 73, Game: 1155 (112) vs 1118 (39)
Season 1988: Highest score difference = 68, Game: 1328 (152) vs 1147 (84)
Season 1989: Highest score difference = 70, Game: 1242 (115) vs 1135 (45)
Season 1990: Highest score difference = 76, Game: 1181 (130) vs 1217 (54)
Season 1991: Highest score difference = 68, Game: 1163 (115) vs 1148 (47)
Season 1992: Highest score difference = 82, Game: 1116 (128) vs 1126 (46)
Season 1993: Highest score difference = 81, Game: 1328 (146) vs 1197 (65)
Season 1994: Highest score difference = 69, Game: 1228 (121) vs 1152 (52)
Season 1995: Highest score difference = 74, Game: 1246 (124) vs 1404 (50)
Season 1996: Highest score difference = 91, Game: 1409 (141) vs 1341 (50)
Season 1997: Highest score difference = 80, Game: 1278 (

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [68]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [69]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [70]:
type(df['Wscore'])

pandas.core.series.Series

In [71]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [72]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

Here's an equivalent using iloc

In [43]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [73]:
df.isnull().sum()

Season             0
Daynum             0
Wteam              0
Wscore             0
Lteam              0
Lscore             0
Wloc               0
Numot              0
ScoreDifference    0
HighScoringGame    0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y